File: python/problems/problem169.py
"""Problem 169 - Project Euler

Exploring the number of different ways a number can be expressed as a sum of powers of 2

Define f(0)=1 and f(n) to be the number of different ways n can be expressed as a sum of integer powers
of 2 using each power no more than twice.

For example, f(10)=5 since there are five different ways to express 10:

  | 1 + 1 + 8
  | 1 + 1 + 4 + 4
  | 1 + 1 + 2 + 2 + 4
  | 2 + 4 + 4
  | 2 + 8

What is f(10^25)?

---------------------------------------------------------------------------------------------------
Optimizations
 * 2^x < 10^25  => x = 83

find the solution with fewest values then use math to decompose those arguments
 | 10: [3,1]
 | 10^25 : [83, 78, 74, 72, 71, 68, 66, 64, 60, 58, 57, 52, 50, 40, 38, 35, 30, 27, 25]

Are there any patterns to this?
f(2^n) = n+1     f(1)=1, f(2)=2, f(4)=3, f(8)=4, ...
f(2^n -1) = 1    f(3)=1, f(7)=1, f(15)=1, ...

Apparently this is a thing: https://oeis.org/A002487, Stern's diatomic series.
a(0) = 0
a(1) = 1
a(2*n) = a(n)
a(2*n+1) = a(n) + a(n+1)
Note, however, that the diatomic series is off by one from f(n):
series:  0, 1, 1, 2, 1, 3, 2, 3, 1, 4, 3, 5, 2, 5, 3, 4, 1, ...
That means, in the series, 2^n=1, 2^n+1=n
"""

# seed cache with powers of 2
cache = {2**i:1 for i in range(0, 25)}


def a(n: int) -> int:
    if n in cache:
        return cache[n]
    k = n // 2
    if n % 2 == 0:
        v = a(k)
    else:
        v = a(k) + a(k+1)
    cache[n] = v
    return v


class Problem169(object):
    @staticmethod
    def solve(n=10**25):
        return a(n+1)

    @staticmethod
    def get_tests():
        return [(10, 5), (1e25, 178653872807)]


if __name__ == '__main__':
    print("The answer is", Problem169.solve())

File: go/euler/internal/problem169.go
package internal

import (
	"math/big"
)

type Problem169 struct {
	big0  *big.Int
	big1  *big.Int
	big2  *big.Int
	cache map[string]int
}

func (p *Problem169) Name() string {
	return "Problem 169"
}

func (p *Problem169) Desc() string {
	return "Exploring the number of different ways a number can be expressed as a sum of powers of 2"
}

func (p *Problem169) Solve() int {
	// from python notes, the diatomic series:
	// a(0)=0, a(1)=1, a(2n)=a(n), a(2n+1)=a(n)+a(n+1)
	// 0, 1, 1, 2, 1, 3, 2, 3, 1, 4, 3, 5, ...
	bn := big.NewInt(10)
	bn = bn.Exp(bn, big.NewInt(25), nil)

	p.big0 = big.NewInt(0)
	p.big1 = big.NewInt(1)
	p.big2 = big.NewInt(2)
	// seed the cache with powers of 2 (which have exactly 1 way)
	p.cache = make(map[string]int)
	for i := 0; i < 25; i++ {
		e := big.NewInt(2)
		e = e.Exp(e, big.NewInt(int64(i)), nil)
		p.cache[e.String()] = 1
	}

	// NOTE, there is an offset from that series and our problem, so add 1 to input
	bn = bn.Add(bn, p.big1)
	return p.a(bn)
}

func (p *Problem169) Run() {
	runProblem(p, 178653872807)
}

// recursive, memoizing helper
func (p *Problem169) a(n *big.Int) int {
	key := n.String()
	v, found := p.cache[key]
	if !found {
		k := big.NewInt(0)
		m := big.NewInt(0)
		k, m = k.DivMod(n, p.big2, m)
		if m.Cmp(p.big0) == 0 {
			v = p.a(k)
		} else {
			v = p.a(k)
			k = k.Add(k, p.big1)
			v += p.a(k)
		}
		p.cache[key] = v
	}
	return v
}
